src/lib/db/public-signup-token-store.ts

```diff
@@ -0,0 +1,189 @@
+        this.timer = (action: string) =>
+            metricsHelper.wrapTimer(eventBus, DB_TIME, {
+                store: 'public-signup-tokens',
+                action,
+            });
+    }
+
+    count(): Promise<number> {
+        return this.db(TABLE)
+            .count('*')
+            .then((res) => Number(res[0].count));
+    }
+
+    private makeTokenUsersQuery() {
+        return this.db<ITokenRow>(`${TABLE} as tokens`)
+            .leftJoin(
+                `${TOKEN_USERS_TABLE} as token_project_users`,
+                'tokens.secret',
+                'token_project_users.secret',
+            )
+            .leftJoin(`users`, 'token_project_users.user_id', 'users.id')
+            .select(
+                'tokens.secret',
+                'tokens.name',
+                'tokens.expires_at',
+                'tokens.created_at',
+                'tokens.created_by',
+                'tokens.role_id',
+                'token_project_users.user_id',
+                'users.name',
+            );
+    }
+
+    async getAll(): Promise<PublicSignupTokenSchema[]> {
+        const stopTimer = this.timer('getAll');
+        const rows = await this.makeTokenUsersQuery();
+        stopTimer();
+        return toTokens(rows);
+    }
+
+    async getAllActive(): Promise<PublicSignupTokenSchema[]> {
+        const stopTimer = this.timer('getAllActive');
+        const rows = await this.makeTokenUsersQuery()
+            .where('expires_at', 'IS', null)
+            .orWhere('expires_at', '>', 'now()');
+        stopTimer();
+        return toTokens(rows);
+    }
+
+    async addTokenUser(secret: string, userId: number): Promise<void> {
+        await this.db.transaction(async (tx) => {
+            await tx<ITokenUserRow>(TOKEN_USERS_TABLE).insert(
+                { user_id: userId, secret },
+                ['created_at'],
+            );
+        });
+    }
+
+    async insert(
+        newToken: IPublicSignupTokenCreate,
+    ): Promise<PublicSignupTokenSchema> {
+        const response = await this.db.transaction(async (tx) => {
+            const [row] = await tx<ITokenRow>(TABLE).insert(toRow(newToken), [
+                'created_at',
+            ]);
+            return {
+                ...newToken,
+                createdAt: row.created_at,
+                users: [],
+            };
+        });
+        return toTokens([response])[0];
+    }
+
+    destroy(): void {}
+
+    async exists(secret: string): Promise<boolean> {
+        const result = await this.db.raw(
+            `SELECT EXISTS (SELECT 1 FROM ${TABLE} WHERE secret = ?) AS present`,
+            [secret],
+        );
+        const { present } = result.rows[0];
+        return present;
+    }
+
+    async get(key: string): Promise<PublicSignupTokenSchema> {
+        const row = await this.makeTokenUsersQuery()
+            .where('secret', key)
+            .first();
+        return toTokens([row])[0];
```

Comment:
first is an optional, I don't think we're always guaranteed a row here.

Reply:
True, fixed